packages/pay/src/connection.ts

```diff
@@ -0,0 +1,292 @@
+} from 'ilp-packet'
+import {
+  generateFulfillment,
+  generateFulfillmentKey,
+  generatePskEncryptionKey,
+  generateRandomCondition,
+  hash
+} from 'ilp-protocol-stream/dist/src/crypto'
+import {
+  ConnectionCloseFrame,
+  ErrorCode,
+  Frame,
+  IlpPacketType,
+  Packet
+} from 'ilp-protocol-stream/dist/src/packet'
+import { Plugin } from 'ilp-protocol-stream/dist/src/util/plugin-interface'
+import { PaymentError } from '.'
+import {
+  SendState,
+  ControllerMap,
+  StreamReject,
+  StreamReply,
+  StreamRequest,
+  StreamRequestBuilder,
+  isFulfillable
+} from './controllers'
+import { FailureController } from './controllers/failure'
+import { PendingRequestTracker } from './controllers/pending-requests'
+import { SequenceController } from './controllers/sequence'
+import {
+  getConnectionId,
+  Integer,
+  timeout,
+  toBigNumber,
+  toLong,
+  getDefaultExpiry,
+  APPLICATION_ERROR_REJECT,
+  MIN_MESSAGE_WINDOW,
+  createReject,
+  ILP_ERROR_CODES
+} from './utils'
+
+/** Serialize & send, and receive & authenticate all ILP and STREAM packets */
+export interface StreamConnection {
+  /** Send packets as frequently as controllers will allow, until they end the send loop */
+  runSendLoop(): Promise<SendState | PaymentError>
+  /** Send an ILP Prepare over STREAM, then parse and validate the reply */
+  sendRequest(request: StreamRequest): Promise<StreamReply | StreamReject>
+  /** Send a `ConnectionClose` frame and disconnect the plugin */
+  close(): Promise<void>
+  /** Logger namespaced to this STREAM connection */
+  log: Logger
+}
+
+export const createConnection = async (
+  plugin: Plugin,
+  controllers: ControllerMap,
+  sharedSecret: Buffer,
+  destinationAddress: string,
+  getExpiry = getDefaultExpiry
+): Promise<StreamConnection> => {
+  const log = createLogger(`ilp-pay:${getConnectionId(destinationAddress)}`)
+
+  const encryptionKey = await generatePskEncryptionKey(sharedSecret)
+  const fulfillmentKey = await generateFulfillmentKey(sharedSecret)
+
+  // Reject all incoming packets, but ACK incoming STREAM packets and handle connection closes
+  plugin.deregisterDataHandler()
+  plugin.registerDataHandler(async data => {
+    try {
+      const prepare = deserializeIlpPrepare(data)
+      const streamRequest = await Packet.decryptAndDeserialize(encryptionKey, prepare.data)
+
+      if (streamRequest.ilpPacketType !== IlpPacketType.Prepare) {
+        return APPLICATION_ERROR_REJECT
+      }
+
+      log.warn(
+        'got incoming Prepare for %s. rejecting with F99: cannot receive money or data. sequence: %s',
+        prepare.amount,
+        streamRequest.sequence
+      )
+      log.trace('got Prepare with frames: %j', streamRequest.frames)
+
+      // In case the server closed the stream/connection, end the payment
+      controllers.get(FailureController).handleRemoteClose(streamRequest.frames, log)
+
+      // No frames are necessary, since on connect we told the receive we can't receive money or data
```

Comment:
Here's where it sends a few frames to tell the receiver it can't receive money or data:

https://github.com/interledgerjs/interledgerjs/blob/dc4300ba33b4ec2b042170b5e5c74912301daa15/packages/pay/src/controllers/asset-details.ts#L68-L85